doc/icmd.8
.TH ICMD 8 "Version 1.0: 10 Aug 2004"
.SH NAME
ipmiutil_cmd \- a tool to send specific IPMI commands via the command line.

.SH SYNOPSIS
.B "ipmiutil cmd [-pqsx -NUPREFJTVY] bus rsSa netFn/lun cmd [data bytes]"

.SH DESCRIPTION
This
.I ipmiutil cmd
tool sends specific IPMI commands to the firmware.  The commands
are composed as hex values on the command line.  This tool was
written to allow in-band use to match the DOS CMDTOOL.EXE or
IPMITOOL.EXE program which is distributed with many Intel servers.
Certain scripts or pre-written commands may have been supplied for
the DOS tool that can now be used while the system is running
Linux or Windows.

This utility can use either the /dev/ipmi0 driver from OpenIPMI,
the /dev/imb driver from Intel, the /dev/ipmikcs driver from valinux,
direct user-space IOs, or the IPMI LAN interface if \-N.

This tool should only be used if you are familiar with the IPMI 1.5
specification, or you have specific pre-written commands to send.


.SH OPTIONS
Command line options are described below.
.IP "-p"
Port to use.  Defaults to RMCP port 623.
.IP "-q"
Quiet mode.  Show only minimal header information.
.IP "-s"
Skips the GetDeviceID command
.IP "-x"
Causes extra debug messages to be displayed.
.IP "-N nodename"
Nodename or IP address of the remote target system.  If a nodename is
specified, IPMI LAN interface is used.  Otherwise the local system
management interface is used.
.IP "-U rmt_user"
Remote username for the nodename given.  The default is a null username.
.IP "-P/-R rmt_pswd"
Remote password for the nodename given.  The default is a null password.
.IP "-E"
Use the remote password from Environment variable IPMI_PASSWORD.
.IP "-F drv_t"
Force the driver type to one of the followng:
imb, va, open, gnu, landesk, lan, lan2, lan2i, kcs, smb.
Note that lan2i means lan2 with intelplus.
The default is to detect any available driver type and use it.
.IP "-J"
Use the specified LanPlus cipher suite (0 thru 17): 0=none/none/none,
1=sha1/none/none, 2=sha1/sha1/none, 3=sha1/sha1/cbc128, 4=sha1/sha1/xrc4_128,
5=sha1/sha1/xrc4_40, 6=md5/none/none, ... 14=md5/md5/xrc4_40.
Default is 3.
.IP "-T"
Use a specified IPMI LAN Authentication Type: 0=None, 1=MD2, 2=MD5, 4=Straight Password, 5=OEM.
.IP "-V"
Use a specified IPMI LAN privilege level. 1=Callback level, 2=User level, 3=Operator level, 4=Administrator level (default), 5=OEM level.
.IP "-Y"
Yes, do prompt the user for the IPMI LAN remote password.
Alternatives for the password are \-E or \-P.

.SH PARAMETERS
The following parameters are used by icmd.  Each is represented as
a two-digit hex byte.  The parameters have no default values.

.IP "bus"
This byte contains the bus number for this command, usually 00.

.IP "rsSa"
This is the resource slave address, usually 0x20 for the BMC.

.IP "netFn/lun"
This byte combines the net Function and Lun.  The 2 low-order bits
are the Lun and the 6 high-order bits are the net Function.
This representation is consistent with the DOS CMDTOOL/IPMITOOL.

.IP "cmd"
This byte contains the IPMI command.

.IP "[data bytes]"
This is a sequence of zero to 16 bytes that represent data bytes
specific to this command.

.SH "EXAMPLES"
icmd 00 20 18 01
.br
Sends the GetDevice ID command to the BMC.
.PP
icmd 00 20 28 43 00 00 ff ff 00 ff
.br
Sends a Get SEL entry command for the last entry in the firmware log.

.SH "SEE ALSO"
ipmiutil(8) ialarms(8) iconfig(8) idiscover(8) ievents(8) ifru(8) igetevent(8) ihealth(8) ilan(8) ireset(8) isel(8) isensor(8) iserial(8) isol(8) iwdt(8)

.SH WARNINGS
See http://ipmiutil.sourceforge.net/ for the latest version of ipmiutil and any bug fix list.

.SH COPYRIGHT
Copyright (C) 2009  Kontron America, Inc.
.PP
See the file COPYING in the distribution for more details
regarding redistribution.
.PP
This utility is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY.

.SH AUTHOR
.PP
Andy Cress <arcress at users.sourceforge.net>
.br



NEWS
IPMIUTIL NEWS 

In October 2001, this project started as part of the Carrier Grade Linux effort.
It was then known as 'panicsel' ("http://panicsel.sf.net") and included a kernel patch to write a Linux panic event to the IPMI firmware log, as well as utilities.  

In July 2002, this code was first included in MontaVista CGE Linux 2.1.
The panicsel functionality was included in OSDL CGL 1.0 and 2.0 requirements.
The kernel panic functionality was included in the OpenIPMI driver for 2.6 Linux kernels, and in the Intel IMB Linux driver v28 and greater.

A Linux rpm and a Windows setup package for ipmiutil is included on 
the platform CD with Intel carrier-grade servers.

The ipmiutil (or panicsel) rpm is known to be included in the following 
Linux distributions:
MontaVista CGE 2.1, 3.0, 3.1, 4.0, SuSE SLES9, SLES10, RedFlag 5.0

See http://ipmiutil.sourceforge.net for binaries, documentation, etc.

In Oct 2001 (0.8.1), started panicsel.
In Dec 2001 (1.0.0), released with Intel Telecom Linux Technology project
In Jul 2002 (1.2.1), released with MontaVista CGE 2.1
In Jan 2004 (1.4.1), added support for Windows 2000 & 2003
In Aug 2004 (1.4.8), the project was moved from panicsel.sf.net to ipmiutil.sf.net.  The new name more clearly reflects the purpose of the project in its current state.  
In Nov 2004 (1.5.0), support for IPMI LAN was added using FreeIPMI.
In May 2006 (1.7.0), built-in IPMI LAN support was added.
In Aug 2006 (1.7.5), driverless direct I/O support was added
In Jan 2007 (1.8.0), added ipmiutil meta-command, IPMI LAN 2.0 (lanplus) support
In Feb 2007 (1.9.2), added Linux Serial-Over-Lan console feature 
In Oct 2007 (2.0.0), added Windows SOL console and remote soft-shutdown features
In Mar 2008 (2.1.0), the Windows SOL client application was completed.
In Aug 2008 (2.2.0), iconfig function was added to save/restore the BMC configuration parameters
In Sep 2008 (2.2.1), support for the Microsoft IPMI driver was added.
In Nov 2008 (2.3.0), support for the Sun Solaris bmc driver was added.
In Mar 2009 (2.3.5), support for the FreeBSD OS was added.
In Oct 2009 (2.5.0), added ANSI VT100  emulation for Windows SOL console
In Feb 2010 (2.6.0), new subcommand naming scheme (i+subcmd)
In Apr 2010 (2.6.2), ifruset was added to set all FRU Product Area fields
In Jun 2010 (2.6.5), added picmg and firewall functionality
In Aug 2010 (2.6.8), Fedora 14 version of ipmiutil submitted
In Sep 2010 (2.6.9), added fwum/hpm functionality, and OEM Fujitsu sensor/sel 
In Sep 2010 (2.7.0), added sunoen, ekanalyzer, fru_picmg, oem_kontron
In Sep 2010 (2.7.1), added ppc64, mips64, sparc arch support
In Nov 2010 (2.7.2), added Debian package support
In Nov 2010 Fedora 14 released with ipmiutil-2.6.8
In Dec 2010 (2.7.3), added SuperMicro OEM functions
In Mar 2011 OpenSuSE 11.4 released with ipmiutil-2.7.2
In May 2011 (2.7.6), added VLAN support and WIN64 fixes
In Sep 2011 (2.7.8), added IPv6, systemd, Dell OEM functions, PICMG 2.3, tsol
In Oct 2011 (2.7.9), support Wind River and ARM cross-compile, Intel S2600
In Dec 2011 (2.8.0), added DCMI 1.1 commands, devel pkg, MSI package
In Feb 2012 (2.8.1), added Quanta S99Q functions, Dell fixes, Intel ME fixes
In Mar 2012 (2.8.2), added Intel Romley OEM support
In Jun 2012 (2.8.4), added support for HP-UX builds and DCMI 1.1 commands
In Oct 2012 (2.8.6), added pre-built Windows DLL for sample development apps
In Feb 2013 (2.8.8), added much more SuperMicro OEM support
In Apr 2013 (2.9.0), added iseltime to sync sys/RTC/SEL times
In May 2013 (2.9.1), added support for MacOSX client
In Dec 2016 (3.0.1), added support for Lenovo OEM sensors
In Feb 2017 (3.0.1), added support for ASUS OEM sensors
In Mar 2017 (3.0.3), added iuser to consolidate user functions



util/icmd.c
/*
 * icmd.c
 *
 * This tool takes command line input as an IPMI command.
 * 
 * Author:  Andy Cress  arcress at users.sourceforge.net
 * Copyright (c) 2003-2006 Intel Corporation. 
 * Copyright (c) 2009 Kontron America, Inc.
 *
 * 03/30/04 Andy Cress - created
 * 04/08/04 Andy Cress - display response data, changed usage order
 * 05/05/04 Andy Cress - call ipmi_close before exit
 * 11/01/04 Andy Cress - add -N / -R for remote nodes   
 * 12/08/04 Andy Cress v1.5 changed usage order, moved bus to first byte,
 *                       gives better compatibility with ipmicmd.exe.
 * 04/13/06 Andy Cress v1.6 fixed istart for -U -R
 */
/*M*
Copyright (c) 2006, Intel Corporation
All rights reserved.

Redistribution and use in source and binary forms, with or without 
modification, are permitted provided that the following conditions are met:

  a.. Redistributions of source code must retain the above copyright notice, 
      this list of conditions and the following disclaimer. 
  b.. Redistributions in binary form must reproduce the above copyright notice,
      this list of conditions and the following disclaimer in the documentation 
      and/or other materials provided with the distribution. 
  c.. Neither the name of Intel Corporation nor the names of its contributors 
      may be used to endorse or promote products derived from this software 
      without specific prior written permission. 

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND 
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED 
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR 
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES 
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON 
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT 
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS 
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *M*/
#if defined(EFI)
 #ifndef NULL
    #define NULL    0
 #endif
 #include <types.h>
 #include <libdbg.h>
 #include <unistd.h>
 #include <errno.h>
#else
 /* Linux, Solaris, BSD, Windows */
 #include <stdio.h>
 #include <stdlib.h>
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <fcntl.h>
 #include <string.h>
 #ifdef WIN32
  #include "getopt.h"
 #elif defined(DOS)
  #include <dos.h>
  #include "getopt.h"
 #elif defined(HPUX)
  /* getopt is defined in stdio.h */
#elif defined(MACOS)
/* getopt is defined in unistd.h */
#include <unistd.h>
 #else
  #include <getopt.h>
 #endif
#endif
#include "ipmicmd.h"
extern void ipmi_lan_set_timeout(int ipmito, int tries, int pingto); 
 
/*
 * Global variables 
 */
#ifdef METACOMMAND
extern char * progver;  /*from ipmiutil.c*/
static char * progname  = "ipmiutil cmd";
#else
static char * progver   = "3.08";
static char * progname  = "icmd";
#endif
static char   fdebug    = 0;
static char   fquiet    = 0;
static char   fset_mc   = 0;
static char   ftest     = 0;
static int    vend_id;
static int    prod_id;
static char   fmBMC = 0;
static char   fdecimal = 0; 
extern int   fjustpass;  /*see ipmicmd.c*/

#define PRIVATE_BUS_ID      0x03 // w Sahalee,  the 8574 is on Private Bus 1
#define PERIPHERAL_BUS_ID   0x24 // w mBMC, the 8574 is on the Peripheral Bus
#define MAXRQLEN 64
#define MAXRSLEN 200

static char usagemsg[] = "Usage: %s [-kqmsxEFNPRTUV] bus rsSa netFn/lun cmd [data bytes]\n";
static uchar busid = PRIVATE_BUS_ID;
static uchar g_bus = PUBLIC_BUS;
static uchar g_sa  = 0;    /*0x20*/
static uchar g_lun = BMC_LUN;
static uchar g_addrtype = ADDR_SMI;

static int send_icmd(uchar *cmd, int len)
{
	uchar responseData[MAXRSLEN];
	int responseLength;
	uchar completionCode;
	int ret, i;
	uchar netfn, lun;
        ushort icmd;

        /* icmd format: 0=bus, 1=rsSa, 2=netFn/lun, 3=cmd,  4-n=data */
	netfn = cmd[2] >> 2;
	lun = cmd[2] & 0x03;
        icmd = cmd[3] + (netfn << 8);
        if (fjustpass) {
           ret = ipmi_cmdraw(IPMB_SEND_MESSAGE,NETFN_APP,BMC_SA,0,0,
			&cmd[0],(uchar)(len), responseData, &responseLength, 
			&completionCode, fdebug);
        } else {
	   responseLength = sizeof(responseData);
           if (g_addrtype == ADDR_IPMB) {  /* && (g_sa != 0) */
              /* if -m used, do ipmi_cmd_mc for IPMB commands  */
              ipmi_set_mc(g_bus,g_sa,g_lun,g_addrtype);
              ret = ipmi_cmd_mc(icmd, &cmd[4], (uchar)(len-4), responseData, 
				&responseLength, &completionCode, fdebug);
              ipmi_restore_mc();
           } else {
	      /* ipmi_cmdraw: bus, cmd, netfn, sa, lun, pdata, ... */
              ret = ipmi_cmdraw(cmd[3], netfn, cmd[1], cmd[0], lun, &cmd[4], 
			(uchar)(len-4), responseData, &responseLength, 
			&completionCode, fdebug);
           }
        }
	if (ret < 0) {
           printf("ipmi_cmd: ret = %d %s\n",ret,decode_rv(ret));
	   return(ret);
        } else if ((ret != 0) || (completionCode != 0)) {
           printf("ipmi_cmd: ret = %d, ccode %02x %s\n",
                    ret, completionCode, decode_cc(icmd,completionCode));
	   return(ret);
	} else if (responseLength > 0) {
	   /* show the response data */
	   printf("respData[len=%d]: ",responseLength);
	   for (i = 0; i < responseLength; i++)
	      printf("%02x ",responseData[i]);
	   printf("\n");
	}
	return(ret);
}  /*end send_icmd()*/


#ifdef METACOMMAND
int i_cmd(int argc, char **argv)
#else
#ifdef WIN32
int __cdecl
#else
int
#endif
main(int argc, char **argv)
#endif
{
   int ret = 0;
   int c;
   uchar devrec[17];
   uchar cmdbuf[MAXRQLEN];
   int maxlen = MAXRQLEN;
   int i, j, istart, cmdlen;
   int fskipdevid = 0;
   int fprivset = 0;
   int cmdmin;
   char *s1;

   istart = 1;
   while ( (c = getopt( argc, argv,"djkm:qst:xp:N:P:R:U:EF:J:T:V:YZ:?")) != EOF ) 
      switch(c) {
          case 'j': /* just pass the bytes to KCS */
		    fjustpass = 1;
		    break;
          case 'd': /* decimal command bytes, skipping bus and sa, 
		       which matches 'ipmitool raw' format. */
		    fdecimal = 1;
		    break;
          case 'm': /* specific MC, 3-byte address, e.g. "409600" */
                    g_bus = htoi(&optarg[0]);  /*bus/channel*/
                    g_sa  = htoi(&optarg[2]);  /*device slave address*/
                    g_lun = htoi(&optarg[4]);  /*LUN*/
                    if (optarg[6] == 's') {
                             g_addrtype = ADDR_SMI;  s1 = "SMI";
                    } else { g_addrtype = ADDR_IPMB; s1 = "IPMB"; }
		    fset_mc = 1;
                    printf("set MC at %s bus=%x sa=%x lun=%x\n",
                            s1,g_bus,g_sa,g_lun);
                    break;
          case 'q': fquiet = 1; fskipdevid = 1; break; /* minimal output */
          case 's': fskipdevid = 1;   break;  /* skip devid */
          case 't': /* set IPMI timeout for ipmilan, usu 2 * 4 = 8 sec */
		    i = atoi(optarg);
		    if (i >= 2) { j = i / 2; i = j; }
		    else { i = 1; j = 1; }
		    ipmi_lan_set_timeout(i,j,1);
		    break;
          case 'k': /* check for IPMI access */
		    ftest = 1;
		    break;
          case 'x': fdebug = 1;       break;  /* debug messages */
          case 'V':    /* priv level */
		fprivset = 1;
          case 'p':    /* port */
          case 'N':    /* nodename */
          case 'U':    /* remote username */
          case 'P':    /* remote password */
          case 'R':    /* remote password */
          case 'E':    /* get password from IPMI_PASSWORD environment var */
          case 'F':    /* force driver type */
          case 'T':    /* auth type */
          case 'J':    /* cipher suite */ 
          case 'Y':    /* prompt for remote password */
          case 'Z':    /* set local MC address */
                parse_lan_options(c,optarg,fdebug);
                break;
	  default:
		printf("%s ver %s\n", progname,progver);
                printf(usagemsg, progname);
                printf(" where -x   shows eXtra debug messages\n"); 
                printf("       -d   decimal input, matching ipmitool syntax\n");
                printf("       -k   check for IPMI access\n"); 
                printf("       -m002000 specific MC (bus 00,sa 20,lun 00)\n"); 
                printf("       -q   quiet mode, with minimal headers\n"); 
                printf("       -s   skips the GetDeviceID command\n"); 
		print_lan_opt_usage(1);
                ret = ERR_USAGE;
		goto do_exit;
      }

   if (ftest) {
	ret = ipmi_getdeviceid(devrec,16,fdebug);
	if (ret == 0) {
	   /*check if a driver is loaded, or direct*/
	   i = get_driver_type();
	   printf("IPMI access is ok, driver type = %s\n",show_driver_type(i));
	   if ((i ==  DRV_KCS) || (i == DRV_SMB)) 
	      printf("Using driverless method\n");
	}
	else printf("IPMI access error %d\n",ret);
	goto do_exit;
   }

   if (!fquiet) {
      printf("%s ver %s\n", progname,progver);
      printf("This is a test tool to compose IPMI commands.\n");
      printf("Do not use without knowledge of the IPMI specification.\n");
   }
   istart = optind;
   if (fdebug) printf("icmd: argc=%d istart=%d\n",argc,istart);

   if (fset_mc == 1) {
      /* target a specific MC via IPMB (usu a picmg blade) */
      ipmi_set_mc(g_bus,g_sa,g_lun,g_addrtype);
   } else g_sa = BMC_SA;

   for (i = 0; i < istart; i++) {
      argv++; argc--;
   }
   if (argc < maxlen) maxlen = argc;
   if (fdebug) printf("icmd: len=%d, cmd byte0=%s\n",maxlen,argv[0]);
   if (fdecimal) {
      uchar b;
      cmdbuf[0] = g_bus;
      cmdbuf[1] = g_sa;
      for (i = 0; i < maxlen; i++) {
	 if (argv[i][0] == '0' && argv[i][1] == 'x')  /*0x00*/
              b = htoi(&argv[i][2]);
         else b = atob(argv[i]);  /*decimal default, or 0x00*/
	 if (i == 0) { /*special handling for netfn */
	    cmdbuf[i+2] = (b << 2) | (g_lun & 0x03);
	 } else 
            cmdbuf[i+2] = b;
      }
      cmdlen = i + 2;
   } else {
      for (i = 0; i < maxlen; i++) {
         cmdbuf[i] = htoi(argv[i]);
      }
      cmdlen = i;
   } 

   if (fdebug) {
        printf("ipmi_cmd: ");
	for (i = 0; i < cmdlen; i++)
	    printf("%02x ",cmdbuf[i]);
        printf("\n");
   }

   if (is_remote() && fprivset == 0) { /*IPMI LAN, privilege not set by user*/
       /* commands to other MCs require admin privilege */
       // if ((g_sa != BMC_SA) || (cmdbuf[1] != BMC_SA))
	   parse_lan_options('V',"4",0);
   }

   if (!fskipdevid) {
    /* 
     * Check the Device ID to determine which bus id to use.
     */
    ret = ipmi_getdeviceid(devrec,16,fdebug);
    if (ret != 0) {
	goto do_exit;
    } else {
      uchar b, j;
      char *pstr;

      if (fdebug) {
	printf("devid: ");
	for (j = 0; j < 16; j++) printf("%02x ",devrec[j]);
	printf("\n");
      }
      b = devrec[4] & 0x0f;
      j = devrec[4] >> 4;
      prod_id = devrec[9] + (devrec[10] << 8);
      vend_id = devrec[6] + (devrec[7] << 8) + (devrec[8] << 16);
      if (vend_id == VENDOR_NSC) {         /* NSC = 0x000322 */
	  fmBMC = 1;  /*NSC miniBMC*/
      } else if (vend_id == VENDOR_INTEL) {  /* Intel = 0x000157 */
         switch(prod_id) {
            case 0x4311:  /* Intel NSI2U*/
 	      fmBMC = 1;  /* Intel miniBMC*/
              break;
            case 0x003E:  /* NSN2U or  CG2100*/
	      set_max_kcs_loops(URNLOOPS); /*longer for cmds (default 300)*/
              break;
	    default:
              break;
	 }
      }
      if (fmBMC) {  /*NSC mini-BMC*/
	  pstr = "mBMC";
	  busid = PERIPHERAL_BUS_ID;  /*used by alarms MWR*/
      } else {  /* treat like Intel Sahalee = 57 01 */
	  pstr = "BMC";
	  busid = PRIVATE_BUS_ID;     /*used by alarms MWR*/
      }
      printf("-- %s version %x.%x, IPMI version %d.%d \n",
             pstr, devrec[2],  devrec[3], b, j);
    }
    // ret = ipmi_getpicmg( devrec, sizeof(devrec),fdebug);
    // if (ret == 0) fpicmg = 1;
   } /*endif skip devid*/

   if (fjustpass) cmdmin = 2;
   else cmdmin = 4;
   if (cmdlen < cmdmin) {
	printf("command length (%d) is too short\n",cmdlen);      
	printf(usagemsg, progname);
   } else {
	ret = send_icmd(cmdbuf,cmdlen); 
        if (!fquiet) printf("send_icmd ret = %d\n",ret); 
   }
do_exit:
   ipmi_close_();
   // if (!fquiet) show_outcome(progname,ret); 
   return (ret);
}  /* end main()*/

/* end icmd.c */
